PyPi release

jannisborn opened the pull request:
Major PR summarising changes that will turn `paperscraper` into a `pypi` package.

Contributions:

- integration of chemrxiv API: Entire dump is downloaded via the code in `paperscraper.get_dumps.utils.chemrxiv` and converted to the same format as the biorxiv and medrxiv dumps, so that `XRXivQuery` can be used.
- migration of the dump-downloads from `bin` into the package to have it standalone. (`paperscraper.get_dumps.biorxiv` etc)
- integration of `scholarly` package to fetch papers from google scholar (no Boolean operators can be used, but citation counts can be retrieved instead).
- Installation instructions and Tutorial in README
- CI setup (extensive tests are lacking)
- renaming `paper_scraper` to `paperscraper`

drugilsberg reviewed (commented):
I did only slight adjustments to the docs and added the automated upload on PYPI for every commit on `master` that is updating `__version__` in `paperscraper/__init__.py`.
Just merging this PR should upload version 0.0.1.keep in mind we can't delete from PYPI)
There are some issues to keep in mind for future releases:

- the HERE paths are very convoluted and depending on the OS/environment might file. `pkg_resources` and package data should be used for this.
- `server_dumps` and `keyword_dumps` are installed alongside the package by `setup.py`, is this something you want? In this way users will have troubles getting the dumps where they want (but maybe this is a design choice). What would make sense is to move them inside paperscraper to avoid pollution of the site-packages directory of the users.
- docstrings and typing are a bit all over the place in terms of consistency.

drugilsberg reviewed (approved):
See the comment above.

jannisborn commented:
Thanks for the changes and in particular the pypi setup. From my side, I think it's ready to merge @drugilsberg can you have a last brief look at the latest changes?

- I moved `server_dumps` inside `paperscraper` and changed to `pkg_resources`. I think the dumps should be hidden to the users. 
- `keyword_dumps` is removed and does not exist anymore. The user can give a directory path to store the dumps and if it does not exist, it is safely created.
- Fixed most of the docstrings and typing I could find. The `ChemrxivAPI` class in `paperscraper.get_dumps.utils.chemrxiv` is not my own code, so I leave this as is.
- agree, tests should be added in the future
- Some README improvements
- plots can now be saved at any received path, not just in a dedicated folder

drugilsberg reviewed (changes requested):
@jannisborn great job, only one change is required. You have to add `package_data` to the setup call in `setup.py` in order to make sure package data will be able to use/see the folder in case data are there (see this for example.)

C-nit commented on .travis.yml:
```diff
@@ -0,0 +1,23 @@
+  provider: script
+  skip_cleanup: true
+  script: bash .travis/deploy.sh
+  on:
+    branch: master
```
```suggestion
  on:
    tags: true
    branch: master
```
This gives a bit more control over deployment to pypi, as tags need to be pushed explicitly.
Not sure that is in the same spirit as the decision to protect master from direct commits though.

>I implemented the automated upload on PyPI for every commit on master that is updating __version__ in paperscraper/__init__.py (settings here).

@drugilsberg I don't get how is this implemented. Is it just failing the deploy because of an existing version? That would mean only release commits (changing version) will succeed on master.

drugilsberg replied:
That is exactly it, this is what I found in travis and adapted twine to not fail if the version exist.

drugilsberg replied:
No strong feelings change as needed.

C-nit replied:
Ah I see, the skip-existing-dist would probably also prevent the different python jobs from failing.

jannisborn commented:
Thanks a lot both!
I put it as 
```py
...
package_data={'paperscraper.server_dumps': ['*']},
....
```

The structure will be
```
paperscraper/
        __init__.py
        server_dumps/
                ___init__.py
               medrxiv_2020-11-10.jsonl
               ...
setup.py
```
I used the wildcard since the dump names are date-specific.

drugilsberg commented:
> Thanks a lot both!
> I put it as
> 
> ```python
> ...
> package_data={'paperscraper.server_dumps': ['*']},
> ....
> ```
> 
> The structure will be
> 
> ```
> paperscraper/
>         __init__.py
>         server_dumps/
>                 ___init__.py
>                medrxiv_2020-11-10.jsonl
>                ...
> setup.py
> ```
> 
> I used the wildcard since the dump names are date-specific.

Not sure that syntax works, this one is more standard:

```py
package_data={
        'paperscraper': ['server_dumps/*']
```

jannisborn commented:
Merging is blocked due to your change request @drugilsberg.
Strictly speaking we dont need the package_data since the package itself ships without the dumps (dumps are >200MB). When the user executes `paperscraper.get_dumps.biorxiv` the dump is downloaded and then stored in the package root.

There is cornercase where this package_metadata can be useful (`git clone`, downloading the dumps and then install in non-editable mode), so I still included it.

jannisborn commented:
I tested the syntax and it works. Your suggested syntax is OS dependent I thought

drugilsberg reviewed (approved):
Just change the `package_data` to the standard syntax and then it's good to go.
Adapting the Travis CI configuration right now.

C-nit commented:
> Just change the `package_data` to the standard syntax and then it's good to go.

From my understanding the `package_data` is actually not intended here, as the package ships **without** the data. Once users download it with the provided calls, it is found with pkg_resources just fine, right?

jannisborn commented:
`From my understanding the package_data is actually not intended here, as the package ships without the data. Once users download it with the provided calls, it is found with pkg_resources just fine, right?`

Exactly
